src/main/frontend/modules/editor/undo_redo.cljs

```diff
@@ -4,44 +4,69 @@
             [frontend.db.conn :as conn]
             [frontend.modules.datascript-report.core :as db-report]
             [frontend.state :as state]
+            [frontend.util.page :as page-util]
+            [frontend.db.model :as db-model]
             [clojure.set :as set]))
 
 ;;;; APIs
 
-(def ^:private undo-redo-states (atom {}))
+(def undo-redo-states (atom {}))
 (def *pause-listener (atom false))
 
 (defn- get-state
-  []
+  [page]
   (let [repo (state/get-current-repo)]
     (assert (string? repo) "Repo should satisfy string?")
-    (if-let [state (get @undo-redo-states repo)]
-      state
-      (let [new-state {:undo-stack (atom [])
-                       :redo-stack (atom [])}]
-        (swap! undo-redo-states assoc repo new-state)
-        new-state))))
+    (when page
+      (if-let [state (get-in @undo-redo-states [repo page])]
+        state
+        (let [new-state {:undo-stack (atom [])
+                         :redo-stack (atom [])}]
+          (swap! undo-redo-states assoc-in [repo page] new-state)
+          new-state)))))
 
 (defn- get-undo-stack
-  []
-  (-> (get-state) :undo-stack))
+  ([]
+   (get-undo-stack (page-util/get-editing-page-id)))
+  ([page]
+   (-> (get-state page) :undo-stack)))
 
 (defn- get-redo-stack
-  []
-  (-> (get-state) :redo-stack))
+  ([]
+   (get-redo-stack (page-util/get-editing-page-id)))
+  ([page]
+   (-> (get-state page) :redo-stack)))
+
+(defn reset-history
+  [page]
+  (when page
+    (let [repo (state/get-current-repo)]
+      (assert (string? repo) "Repo should satisfy string?")
```

Comment:
Looks like [get-current-repo](https://github.com/logseq/logseq/blob/08d5eaff0976204facb60ba7706d511bf1939d7a/src/main/frontend/state.cljs#L747) always returns a string. Could be more useful to add a `:post` check on that fn instead of external checks. See `frontend.fs.sync` for some good examples

Reply:
Admittedly, I didn't check `get-current-repo` myself. I just copied the assertion from `get-state`, so i removed both.